daprovider/daclient/daclient.go

```diff
@@ -0,0 +1,101 @@
+		ArgLogLimit:               2048,
+		WebsocketMessageSizeLimit: 256 * 1024 * 1024,
+	},
+}
+
+func ClientConfigAddOptions(prefix string, f *pflag.FlagSet) {
+	f.Bool(prefix+".enable", DefaultClientConfig.Enable, "enable daprovider client")
+	f.Bool(prefix+".with-writer", DefaultClientConfig.WithWriter, "implies if the daprovider rpc server supports writer interface")
+	rpcclient.RPCClientAddOptions(prefix+".rpc", f, &DefaultClientConfig.RPC)
+}
+
+func NewClient(ctx context.Context, config rpcclient.ClientConfigFetcher) (*Client, error) {
+	client := &Client{rpcclient.NewRpcClient(config, nil)}
+	if err := client.Start(ctx); err != nil {
+		return nil, fmt.Errorf("error starting daprovider client: %w", err)
+	}
+	return client, nil
+}
+
+// IsValidHeaderByteResult is the result struct that data availability providers should use to respond if the given headerByte corresponds to their DA service
+type IsValidHeaderByteResult struct {
+	IsValid bool `json:"is-valid,omitempty"`
+}
+
+func (c *Client) IsValidHeaderByte(ctx context.Context, headerByte byte) bool {
+	var isValidHeaderByteResult IsValidHeaderByteResult
+	if err := c.CallContext(ctx, &isValidHeaderByteResult, "daprovider_isValidHeaderByte", headerByte); err != nil {
+		log.Error("Error returned from daprovider_isValidHeaderByte rpc method, defaulting to result as false", "err", err)
+		return false
+	}
+	return isValidHeaderByteResult.IsValid
+}
+
+// RecoverPayloadFromBatchResult is the result struct that data availability providers should use to respond with underlying payload and updated preimages map to a RecoverPayloadFromBatch fetch request
+type RecoverPayloadFromBatchResult struct {
+	Payload   hexutil.Bytes           `json:"payload,omitempty"`
+	Preimages daprovider.PreimagesMap `json:"preimages,omitempty"`
+}
+
+func (c *Client) RecoverPayloadFromBatch(
+	ctx context.Context,
+	batchNum uint64,
+	batchBlockHash common.Hash,
+	sequencerMsg []byte,
+	preimages daprovider.PreimagesMap,
+	validateSeqMsg bool,
+) ([]byte, daprovider.PreimagesMap, error) {
+	var recoverPayloadFromBatchResult RecoverPayloadFromBatchResult
+	if err := c.CallContext(ctx, &recoverPayloadFromBatchResult, "daprovider_recoverPayloadFromBatch", hexutil.Uint64(batchNum), batchBlockHash, hexutil.Bytes(sequencerMsg), preimages, validateSeqMsg); err != nil {
+		return nil, nil, fmt.Errorf("error returned from daprovider_recoverPayloadFromBatch rpc method, err: %w", err)
+	}
+	return recoverPayloadFromBatchResult.Payload, recoverPayloadFromBatchResult.Preimages, nil
+}
+
+// StoreResult is the result struct that data availability providers should use to respond with a commitment to a Store request for posting batch data to their DA service
+type StoreResult struct {
+	SerializedDACert hexutil.Bytes `json:"serialized-da-cert,omitempty"`
+}
+
+func (c *Client) Store(
+	ctx context.Context,
+	message []byte,
+	timeout uint64,
+	disableFallbackStoreDataOnChain bool,
```

Comment:
It seems odd to me to have the fallback be decided by provider and not the thing calling the provider.

Reply:
not sure if i understand the question right, the fallback is decided by the caller function though. If this is set to false by the caller function, `Store` would just return the message as is